--- ddclient-3.11.2/ddclient.conf.in.orig	2023-12-11 12:21:48.545194604 +0100
+++ ddclient-3.11.2/ddclient.conf.in	2023-12-11 12:22:41.613657451 +0100
@@ -18,9 +18,9 @@
 ######################################################################
 daemon=300				# check every 300 seconds
 syslog=yes				# log update msgs to syslog
-mail=root				# mail all msgs to root
+#mail=root				# mail all msgs to root
 mail-failure=root		# mail failed update msgs to root
-pid=@runstatedir@/ddclient.pid		# record PID in file.
+pid=@runstatedir@/ddclient/ddclient.pid		# record PID in file.
 ssl=yes					# use ssl-support.  Works with
 						# ssl-library
 # postscript=script		# run script after updating.  The
